spec/prog/vm/github_runner_spec.rb

```diff
@@ -422,6 +422,36 @@
 
       expect { nx.setup_environment }.to hop("register_runner")
     end
+
+    it "hops to register_runner with after enabling transparent cache and using Cloudflare DNS servers with all-servers" do
+      expect(vm).to receive(:vm_host).and_return(instance_double(VmHost, ubid: "vhfdmbbtdz3j3h8hccf8s9wz94", location: "hetzner-fsn1", data_center: "FSN1-DC8", sshable: sshable)).at_least(:once)
+      expect(vm).to receive(:runtime_token).and_return("my_token")
+      expect(github_runner.installation).to receive(:project).and_return(instance_double(Project, ubid: "pjwnadpt27b21p81d7334f11rx", path: "/project/pjwnadpt27b21p81d7334f11rx", get_ff_all_servers_dnsmasq: true)).at_least(:once)
+      expect(github_runner.installation).to receive(:cache_enabled).and_return(true)
+      expect(vm).to receive(:nics).and_return([instance_double(Nic, private_ipv4: NetAddr::IPv4Net.parse("10.0.0.1/32"))])
+      expect(sshable).to receive(:cmd).with(<<~COMMAND)
+        set -ueo pipefail
+        echo "image version: $ImageVersion"
+        sudo usermod -a -G sudo,adm runneradmin
+        jq '. += [{"group":"Ubicloud Managed Runner","detail":"Name: #{github_runner.ubid}\\nLabel: ubicloud-standard-4\\nArch: \\nImage: \\nVM Host: vhfdmbbtdz3j3h8hccf8s9wz94\\nVM Pool: \\nLocation: hetzner-fsn1\\nDatacenter: FSN1-DC8\\nProject: pjwnadpt27b21p81d7334f11rx\\nConsole URL: http://localhost:9292/project/pjwnadpt27b21p81d7334f11rx/github"}]' /imagegeneration/imagedata.json | sudo -u runner tee /home/runner/actions-runner/.setup_info
+        echo "UBICLOUD_RUNTIME_TOKEN=my_token
+        UBICLOUD_CACHE_URL=http://localhost:9292/runtime/github/" | sudo tee -a /etc/environment
+        ([ -f "/etc/docker/daemon.json" ] && sudo cat /etc/docker/daemon.json || echo '{}') | jq '. += #{JSON.generate({"experimental" => false, "dns-opts" => ["timeout:2", "attempts:3"]})}' | sudo tee /etc/docker/daemon.json
+        sudo systemctl restart docker
+        echo "CUSTOM_ACTIONS_CACHE_URL=http://10.0.0.1:51123/random_token/" | sudo tee -a /etc/environment
+      COMMAND
```

Comment:
I might make this string generation a method and test that separately, then you don't need such a complex fixture/mocks, rather than inlining it.

So you might end up with a (very fast, in-memory) test that's like:

```ruby
github_runner.inhost_name = 'foo' # or an expect, or whatever
expect(obj.generate_cloudflare_sed).to eq(<<str)
the string
str
```

Then your test in integration can piggy-back on a different test that does all the fixture stuff, e.g.:

```ruby
expect(sshable).to receive(cmd).with(obj.generate_cloudflare_sed)
```

The net result is intended to be smaller and less intensive to read. As-is, there's a very long string that I take it you need because there are multiple calls to `cmd` and you can't reduce so easily to `receive(:cmd)` without `with`.

An alternative version you may prefer is put even the `cmd` into the helper method. With that approach, you can use a expect/with construction to make sure the string is what you expect, like so:

```ruby
github_runner.inhost_name = 'foo' # or an expect, or whatever
expect(sshable).to receive(:cmd).with('thestring')
obj.run_cloudflare_sed
```

This uglifies relative to a simple `eq`, but the payoff is in the test that integrates `run_cloudflare_sed`:

```ruby
expect(obj).to receive(:run_cloudflare_sed)
```

is all you need.

I realize that was a bit hand-wavey, but does it make sense? the general idea is to eliminate a lot of new-new fixture code and make the tests denser. Currently the ratio of interesting-to-not-very-interesting expressions in this test is rather low.

Reply:
Thank you! I refactored the part that I added but the above CMD is unrelated from this PR. We should fix it in a different patch.